### ISSUE TITLE
[Support]: Accessibility: Keyboard Trap

### ISSUE THREAD

**Issue by philmnort** [ceolter/ag-grid#12547]
### Link to reproducible scenario

https://www.ag-grid.com/example/

### Add a description

# Background
WCAG level A accessibility item 2.1.2 defines "no keyboard trap"
https://www.w3.org/TR/WCAG21/#keyboard-accessible
https://www.w3.org/WAI/WCAG21/Understanding/no-keyboard-trap.html

# Issue
Tab key will tab into the grid, but you cannot tab _out_ of the grid without pressing tab N times, where N is the number of cells in the grid.  This can be in the thousands.  This is effectively a keyboard trap.

# Recommended solution
Tab key focuses _into_ the grid, from there arrow keys are used to change focused cell.  Next tab key press moves focus to next element _after_ the grid.
This is similar to the guidance for menus: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/menu_role

 # Ag-Grid notes
this issue is easily reproducible directly on the ag-grid demo site: https://www.ag-grid.com/example/
Documentation specifically notes tab changes focus _between_ cells, which many would argue is a keyboard trap:
https://www.ag-grid.com/javascript-data-grid/keyboard-navigation/

"If a cell on the first grid row is focused and you press ↑, the focus will be moved into the grid header. The header navigation focus navigation works the same as the grid's: arrows will move up/down/left/right, ⇥ Tab will move the focus horizontally until the last header cell and then move on to the next row."

### Version

all

### What language are you using?

React

### What browser are you using?

Chrome

**Comment by leshields** [comment]
WAI Patterns has examples of data grids that demonstrate the desired tabbing behavior described above: https://www.w3.org/WAI/ARIA/apg/patterns/grid/examples/data-grids/

WAI guidance regarding different types of grid keyboard interactions: https://www.w3.org/WAI/ARIA/apg/patterns/grid/

Note that keyboard focus typically defaults to the first cell in the grid, but after users enter the grid and navigate around, focus is updated to the last visited cell. This allows users to return to the spot they left off at rather than starting at the beginning again if they navigate away from the grid and then back.

**Comment by avramdlavinsky** [comment]
Excellent detail above from leshields. This is a very painful gap. My team has had to implement extensive keyboard navigation adapter files, a messy fix. Please address.

**Comment by kiril-matev** [comment]
Hello @philmnort,

The default behavior of the grid is to move to the next cell when pressing the Tab key. This implements the WCAG requirement correctly. Please note that Excel for Office 365 and Google Sheets have the same behavior as AG Grid - pressing Tab key moves focus to the next cell, not to the next container.

While this complies with WCAG guidance, it doesn't result in a good user experience when you have a grid showing many rows and columns.

You can easily configure the grid to skip directly to the next container when pressing the Tab using the tabToNextCell callback. 

Please see this below:
https://plnkr.co/edit/gHAr45CE2vfshtRi?open=main.js

Focus the button above the grid and press the Tab key multiple times. Note that focus leaves the grid cells in a single Tab. This means that users will only be able to navigate the grid cells using the arrow keys, and use the Tab/Shift+Tab keys to move focus across containers.
